openshift-kube-apiserver/filters/deprecatedapirequest/apiaccess_count_controller.go

```diff
@@ -79,47 +81,83 @@ func (c *controller) Start(stop <-chan struct{}) {
 	}()
 
 	// write out logs every c.updatePeriod
-	go wait.Until(func() {
-		klog.V(2).Infof("updating top APIRequest counts")
-		defer klog.V(2).Infof("finished updating top APIRequest counts")
-
-		// get the current count to persist, start a new in-memory count
-		countsToPersist := c.resetRequestCount()
-
-		// remove stale data
-		expiredHour := (time.Now().Hour() + 1) % 24
-		countsToPersist.ExpireOldestCounts(expiredHour)
-
-		// when this function returns, add any remaining counts back to the total to be retried for update
-		defer c.requestCounts.Add(countsToPersist)
-
-		var wg sync.WaitGroup
-		for _, resourceCount := range countsToPersist.resourceToRequestCount {
-			wg.Add(1)
-			go func(localResourceCount *resourceRequestCounts) {
-				defer wg.Done()
-
-				klog.V(2).Infof("updating top %v APIRequest counts", localResourceCount.resource)
-				defer klog.V(2).Infof("finished updating top %v APIRequest counts", localResourceCount.resource)
-
-				_, _, err := v1helpers.ApplyStatus(
-					ctx,
-					c.client,
-					resourceToAPIName(localResourceCount.resource),
-					SetRequestCountsForNode(c.nodeName, expiredHour, localResourceCount),
-				)
-				if err != nil {
-					runtime.HandleError(err)
-					return
-				}
-
-				// on successful update, remove the counts
-				countsToPersist.RemoveResource(localResourceCount.resource)
-			}(resourceCount)
+	go wait.UntilWithContext(ctx, c.persistRequestCountForAllResources, c.updatePeriod)
+}
+
+func (c *controller) persistRequestCountForAllResources(ctx context.Context) {
+	klog.V(2).Infof("updating top APIRequest counts")
+	defer klog.V(2).Infof("finished updating top APIRequest counts")
+
+	// get the current count to persist, start a new in-memory count
+	countsToPersist := c.resetRequestCount()
+
+	// remove stale data
+	expiredHour := (time.Now().Hour() + 1) % 24
+	currentHour := time.Now().Hour()
+	countsToPersist.ExpireOldestCounts(expiredHour)
+
+	// when this function returns, add any remaining counts back to the total to be retried for update
+	defer c.requestCounts.Add(countsToPersist)
+
+	var wg sync.WaitGroup
+	for gvr := range countsToPersist.resourceToRequestCount {
```

Comment:
so this is for all GVR used in the cluster practically, (including all CRDs), each goroutine does an status update, could (the concurrency) potentially be a bottleneck for a large cluster with many many CRDs?

Reply:
> so this is for all GVR used in the cluster practically, (including all CRDs), each goroutine does an status update, could (the concurrency) potentially be a bottleneck for a large cluster with many many CRDs?

this is during a write phase.  The data collection phase is done on the instance variable specifically to avoid an impact.   This is about 200 resource updates every five minutes.  I wouldn't expect a problem, but there may be one.